builtin/providers/azure/resource_azure_instance.go

```diff
@@ -0,0 +1,653 @@
+}
+
+func retrieveVMImageDetails(
+	mc management.Client,
+	label string) (func(*virtualmachine.Role) error, string, error) {
+	imgs, err := virtualmachineimage.NewClient(mc).ListVirtualMachineImages()
+	if err != nil {
+		return nil, "", fmt.Errorf("Error retrieving image details: %s", err)
+	}
+
+	for _, img := range imgs.VMImages {
+		if img.Label == label {
+			if img.OSDiskConfiguration.OS != linux && img.OSDiskConfiguration.OS != windows {
+				return nil, "", fmt.Errorf("Unsupported image OS: %s", img.OSDiskConfiguration.OS)
+			}
+
+			configureForImage := func(role *virtualmachine.Role) error {
+				return vmutils.ConfigureDeploymentFromVMImage(
+					role,
+					img.Name,
+					"",
+					true,
+				)
+			}
+
+			return configureForImage, img.OSDiskConfiguration.OS, nil
+		}
+	}
+
+	return nil, "", fmt.Errorf("Could not find image with label '%s'", label)
+}
+
+func retrieveOSImageDetails(
+	mc management.Client,
+	label,
+	storage string) (func(*virtualmachine.Role) error, string, error) {
+	imgs, err := osimage.NewClient(mc).ListOSImages()
+	if err != nil {
+		return nil, "", fmt.Errorf("Error retrieving image details: %s", err)
+	}
+
+	for _, img := range imgs.OSImages {
+		if img.Label == label {
+			if img.OS != linux && img.OS != windows {
+				return nil, "", fmt.Errorf("Unsupported image OS: %s", img.OS)
+			}
+			if img.MediaLink == "" {
+				if storage == "" {
+					return nil, "",
+						fmt.Errorf("When using a platform image, the 'storage' parameter is required")
+				}
+				img.MediaLink = fmt.Sprintf(osDiskBlobStorageURL, storage, label)
+			}
+
+			configureForImage := func(role *virtualmachine.Role) error {
+				return vmutils.ConfigureDeploymentFromPlatformImage(
+					role,
+					img.Name,
+					img.MediaLink,
+					label,
+				)
+			}
+
+			return configureForImage, img.OS, nil
+		}
+	}
+
+	return nil, "", fmt.Errorf("Could not find image with label '%s'", label)
+}
+
+func endpointProtocol(p string) virtualmachine.InputEndpointProtocol {
+	if p == "tcp" {
+		return virtualmachine.InputEndpointProtocolTCP
+	}
+
+	return virtualmachine.InputEndpointProtocolUDP
+}
+
+func verifyInstanceParameters(d *schema.ResourceData, osType string) error {
+	if osType == linux {
+		_, pass := d.GetOk("password")
+		_, key := d.GetOk("ssh_key_thumbprint")
+
+		if !pass && !key {
+			return fmt.Errorf(
+				"You must supply a 'password' and/or a 'ssh_key_thumbprint' when using a Linux image")
+		}
+	}
+
+	if osType == windows {
+		if _, ok := d.GetOk("password"); !ok {
+			return fmt.Errorf("You must supply a 'password' when using a Windows image")
+		}
+
+		if _, ok := d.GetOk("time_zone"); !ok {
+			return fmt.Errorf("You must supply a 'time_zone' when using a Windows image")
+		}
+	}
+
+	if _, ok := d.GetOk("subnet"); ok {
+		if _, ok := d.GetOk("virtual_network"); !ok {
+			return fmt.Errorf("You must also supply a 'virtual_network' when supplying a 'subnet'")
```

Comment:
These kinds of parameter dependencies checked directly through the schema would be awesome.
Until then, good job enforcing them here!